### PR TITLE
ROK Connector: Authentication mode with specific JWT

### PR DESCRIPTION
**Pull request by raphaelvignes**
## Change Summary
Implementation of a JWT based authentication method with ROK: 
Live data mode data source now requires a secret (to be provided by ROK) to authenticate users and retrieve data. 
The authentication process is described in the attached document [https://trello-attachments.s3.amazonaws.com/5e1c3818f58aba66fa30fcc0/5f748a3b848ebe1fbd540267/5f63d051d2cac56b5a81aa95033685e2/Proposition_authentification.docx](url)

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable

### CODE REVIEW

File: tests/rok/test_rok.py [hunk 1]
```diff
@@ -1,6 +1,11 @@
 import pytest
 import responses
+<<<<<<< HEAD
+from jwt import decode, encode
 
+=======
+from jwt import encode, decode
+>>>>>>> 1d84ccc84dc24e9f40c8513a84cb4b0a797584ef
```
[the remark this answers is not in the document]
Comment (raphaelvignes) [reply tests/rok/test_rok.py]: Merci pour la revue. 
J'ai corrigé ça.

**Comment by raphaelvignes** [comment]
I'll pause this PR until I have a reply from ROK regarding: 
- JWT format
- Generated Secret

**Comment by raphaelvignes** [comment]
I think, that we can at least review the crafting of token & data retrieval based on mocked API while waiting for ROK feedback on new authentication method.

File: tests/rok/test_rok.py [hunk 2]
```diff
@@ -1,57 +1,129 @@
...
 @responses.activate
-def test_rok_auth_error():
+def test_rok_with_jwt(authentified_con, authentified_ds):
+    """Check that we correctly retrieve the data with the crafted token"""
+    data = {
+        'token': encode(
+            {'rok_token': 'rok_token'}, authentified_con.secret, algorithm='HS256'
+        ).decode('utf-8')
+    }
+    responses.add(responses.POST, endpoint, json={'data': data})
```
[the remark this answers is not in the document]
Comment (davinov) [reply tests/rok/test_rok.py:88]: Pour valider qu'on envoie bien les bons paramètres, on peut utiliser des matchers: https://github.com/getsentry/responses#matching-request-parameters

Aussi: je me demande si rien ne plante si jamais ces réponses ne sont en fait pas appellées.
As-tu fait un test pour vérifier que si on ajotue une réponse qui n'est jamais demandée, alors le test est rouge ?
Si ce n'est pas le cas, je vois qu'il y a un [assert_call_count](https://github.com/getsentry/responses#assert_call_count) dans la lib

File: toucan_connectors/rok/rok_connector.py [hunk 3]
```diff
@@ -14,33 +17,77 @@ class RokDataSource(ToucanDataSource):
...
+                'user': self.username,
+                'password': self.password,
+            }
+            auth_res = requests.post(
+                endpoint, json={'query': auth_query, 'variables': auth_vars}
+            ).json()
+
+            if 'errors' in auth_res:
+                raise ValueError(str(auth_res['errors']))
+            rok_token = auth_res['data']['authenticate']
```
Comment (davinov) [inline toucan_connectors/rok/rok_connector.py:76]: For now, this is not too big, but it would be nice to extract this into an retrieve_token method of the connector, si it can be easily mocked, and keep the logici in retrieve_data easily readable

**Review by davinov — COMMENTED** [review]
Thanks for this work!

(Sorry for the comments in french, I've started the review in fr and now I'm a bit lazy to translate them)

File: doc/connectors/ROK.md [hunk 4]
```diff
@@ -1,6 +1,9 @@
 # ROK connector
 
 Connector to [ROK](https://www.rok-solution.fr/).
+In live data mode, requires a secret to craft a JWT authenticating the user.
+Two authentication methods are available username with password and ROK Token which authenticate the current user. 
+In this later mode, the username must be given as a param in the creation form of the connector.
```
[the remark this answers is not in the document]
Comment (raphaelvignes) [reply doc/connectors/ROK.md:6]: Thanks for this addition, it provides more clarification.

File: toucan_connectors/rok/rok_connector.py [hunk 5]
```diff
@@ -14,33 +17,77 @@ class RokDataSource(ToucanDataSource):
...
+        endpoint = f'{self.host}/graphql'
+
+        if self.authentified_with_rok_token:
+            if not data_source.live_data:
+                raise InvalidAuthenticationMethodError(
+                    'Request with ROK token is not possible while not in live data mode. Change the connector configuration to live data'
+                )
+
+            if not self.secret:
+                raise NoROKSecretAvailableError('secret not defined')
```
[the remark this answers is not in the document]
Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:37]: Thx

File: tests/rok/test_rok.py [hunk 6]
```diff
@@ -1,57 +1,129 @@
...
+
+@pytest.fixture
+def unauthentified_c():
+    return RokConnector(
+        name='RokConnector',
+        host='https://rok.example.com',
+        username='username',
+        password='password',
+    )
+
```
[the remark this answers is not in the document]
Comment (raphaelvignes) [reply tests/rok/test_rok.py:44]: Thx, changed fixtures with more relevant names.

**Comment by raphaelvignes** [comment]
Two more tests need to be performed: 
- Do we get data segregated by users when we use the JWT authentication (can be done locally with quickstart's script) ? 
- Are we able to send dates parameters through interpolation in the query either in JWT & User/Pw authentication mode ? For that I'm waiting confirmation from ROK 

If both are OK, then no modifications are needed for this version of the connector.

File: toucan_connectors/rok/rok_connector.py [hunk 7]
```diff
@@ -12,35 +18,117 @@ class RokDataSource(ToucanDataSource):
...
+
+
 class RokConnector(ToucanConnector):
     data_source_model: RokDataSource
-
     host: str
     username: str
-    password: str
+    password: str = None
+    secret: str = None
```
Comment (raphaelvignes) [inline toucan_connectors/rok/rok_connector.py:41]: I have a doubt on this, shouldn't the secret be retrieved from env variables ?

Comment (davinov) [reply toucan_connectors/rok/rok_connector.py:41]: It will!
We will advise to use some variables to do this, like: `secret: {{ secrets.my_rok_secret_for_db_a }}`

**Comment by raphaelvignes** [comment]
I successfully tested that the data are scoped by users with the JWT authentication. 
However, variable interpolation is not working as-is. I'll add it and as suggested I'll add start_date, end_date & viewid as required parameters in the data_srouce configuration.

**Comment by raphaelvignes** [comment]
PR ready for review:
- Variable interpolation is working
- I also onboarded [https://trello.com/c/jFogyX8w/1095-2-etq-rok-client-linterface-du-connecteur-me-fait-appara%C3%AEtre-clairement-les-param%C3%A8tres-required-dun-requ%C3%AAte-rok](this card) on the PR

File: tests/rok/test_rok.py [hunk 8]
```diff
@@ -283,32 +283,56 @@ def test_live_instance_jwt():
...
+            ['None', 'bar'],
+        ),
+        (
+            {'foo': 'bar'},
+            {'start_date': None, 'end_date': None, 'viewId': None},
+            '{cartographies{dataTracking(startUtcDate:"0", endUtcDate:"1", foo: "2", viewId:"3"{procedureInstancesWithRuleReportings{procedureInstance{id shortDescriptionSetter}}}}}',
+            ['1'],
+        ),
+    ],
+)
```
Comment (davinov) [inline tests/rok/test_rok.py:308]: :+1: Great to parametrize that test :+1:

File: toucan_connectors/rok/rok_connector.py [hunk 9]
```diff
@@ -48,6 +51,20 @@ class RokConnector(ToucanConnector):
...
+        }
+
+        if data_source.parameters:
+            data_source.query = nosql_apply_parameters_to_query(
+                data_source.query, {**data_source.parameters, **date_viewid_parameters}
+            )
+        else:
+            data_source.query = nosql_apply_parameters_to_query(
+                data_source.query, date_viewid_parameters
+            )
```
Comment (davinov) [inline toucan_connectors/rok/rok_connector.py:67]: Nice!
If you want a shortcut for that, you may use:
```suggestion
        if data_source.parameters:
            parameters = {**data_source.parameters, **date_viewid_parameters}
        else:
            parameters = date_viewid_parameters
         data_source.query = nosql_apply_parameters_to_query(
                data_source.query, parameters
            )
```
to centralize the call to the interpolation function

File: toucan_connectors/rok/rok_connector.py [hunk 10]
```diff
@@ -1,12 +1,12 @@
 import base64
 import json
 from datetime import datetime, timedelta
-from json import JSONDecodeError
 
 import pandas as pd
 import requests
 from jwt import encode
 from pydantic import Field
+from simplejson import JSONDecodeError
```
Comment (davinov) [inline toucan_connectors/rok/rok_connector.py:9]: Do you know why we need to import that from the simplejson module ?
Is that related to the jwt module using it internally?

**Review by davinov — COMMENTED** [review]
:100:  Thanks for mastering the tests :heavy_check_mark:

File: toucan_connectors/rok/rok_connector.py (hunk 9 above)
Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:67]: Why make it complicated when it can be as simple as that :). Thanks !

File: toucan_connectors/rok/rok_connector.py (hunk 10 above)
Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:9]: requests seems to be raising simplejson.errors.JSONDecodeError when failling to parse a response object with .json() :

```
self = <simplejson.decoder.JSONDecoder object at 0x7f9fed2bcbb0>, s = '<Fault xmlns= ...', idx = 0, _w = <built-in method match of re.Pattern object at 0x7f9fed9391b0>, _PY3 = True

    def raw_decode(self, s, idx=0, _w=WHITESPACE.match, _PY3=PY3):
        """Decode a JSON document from ``s`` (a ``str`` or ``unicode``
        beginning with a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
        Optionally, ``idx`` can be used to specify an offset in ``s`` where
        the JSON document begins.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        if idx < 0:
            # Ensure that raw_decode bails on negative indexes, the regex
            # would otherwise mask this behavior. #98
            raise JSONDecodeError('Expecting value', s, idx)
        if _PY3 and not isinstance(s, str):
            raise TypeError("Input string must be text, not bytes")
        # strip UTF-8 bom
        if len(s) > idx:
            ord0 = ord(s[idx])
            if ord0 == 0xfeff:
                idx += 1
            elif ord0 == 0xef and s[idx:idx + 3] == '\xef\xbb\xbf':
                idx += 3
>       return self.scan_once(s, idx=_w(s, idx).end())
E       simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Comment (davinov) [reply toucan_connectors/rok/rok_connector.py:9]: Thanks for clarifying it!

Makes me think that I think the ROK connector depends on the `pyjwt` module, and we should add it here: https://github.com/ToucanToco/toucan-connectors/pull/222/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R31)
I'm surprised we have to indicate simplejson as dependecy here, as it's probably already installed by the package that raises the error. Maybe we can remove it ?

Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:9]: I changed that, let's see if the CI is ok

Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:9]: image

Got that when removing simplejson from required dependicies :(

Comment (raphaelvignes) [reply toucan_connectors/rok/rok_connector.py:9]: simplejson finally added to dependencies.

**Review by davinov — APPROVED** [review]
:tada:
